source/unix/openssl_platform_init.c

```diff
@@ -104,26 +369,46 @@ static int s_resolve_libcrypto_md(enum aws_libcrypto_version version, void *modu
     bool has_awslc_symbols =
         md_new_fn && md_create_fn && md_free_fn && md_destroy_fn && md_init_ex_fn && md_update_fn && md_final_ex_fn;
 
-    if (!has_awslc_symbols && version == AWS_LIBCRYPTO_LC) {
+    if (has_awslc_symbols) {
+        FLOGF("found static aws-lc libcrypto 1.1.1 EVP_MD symbols");
+    } else {
         *(void **)(&md_new_fn) = dlsym(module, "EVP_MD_CTX_new");
         *(void **)(&md_free_fn) = dlsym(module, "EVP_MD_CTX_free");
         *(void **)(&md_init_ex_fn) = dlsym(module, "EVP_DigestInit_ex");
         *(void **)(&md_update_fn) = dlsym(module, "EVP_DigestUpdate");
         *(void **)(&md_final_ex_fn) = dlsym(module, "EVP_DigestFinal_ex");
         if (md_new_fn) {
-            FLOGF("found dynamic libcrypto 1.1.1 EVP_MD symbols");
+            FLOGF("found dynamic aws-lc libcrypto 1.1.1 EVP_MD symbols");
         }
     }
 
-    /* Add the found symbols to the vtable */
-    evp_md_ctx_table.new_fn = md_new_fn;
-    evp_md_ctx_table.free_fn = md_free_fn;
-    evp_md_ctx_table.init_ex_fn = md_init_ex_fn;
-    evp_md_ctx_table.update_fn = md_update_fn;
-    evp_md_ctx_table.final_ex_fn = md_final_ex_fn;
-    g_aws_openssl_evp_md_ctx_table = &evp_md_ctx_table;
+    if (md_new_fn) {
+        /* Add the found symbols to the vtable */
+        evp_md_ctx_table.new_fn = md_new_fn;
+        evp_md_ctx_table.free_fn = md_free_fn;
+        evp_md_ctx_table.init_ex_fn = md_init_ex_fn;
+        evp_md_ctx_table.update_fn = md_update_fn;
+        evp_md_ctx_table.final_ex_fn = md_final_ex_fn;
+        g_aws_openssl_evp_md_ctx_table = &evp_md_ctx_table;
+        return true;
+    }
 
-    return version;
+    return false;
+}
+
+static int s_resolve_libcrypto_md(enum aws_libcrypto_version version, void *module) {
+    switch (version) {
+        case AWS_LIBCRYPTO_LC:
+            return s_resolve_md_lc(module) ? version : AWS_LIBCRYPTO_NONE;
+        case AWS_LIBCRYPTO_1_1_1:
+            return s_resolve_md_111(module) ? version : AWS_LIBCRYPTO_NONE;
+        case AWS_LIBCRYPTO_1_0_2:
+            return s_resolve_hmac_102(module) ? version : AWS_LIBCRYPTO_NONE;
+        case AWS_LIBCRYPTO_NONE:
+            AWS_FATAL_ASSERT(!"Attempted to resolve invalid libcrypto MD API version AWS_LIBCRYPTO_NONE");
+    }
+
+    return AWS_LIBCRYPTO_NONE;
 }
 
 static int s_resolve_libcrypto_symbols(enum aws_libcrypto_version version, void *module) {
```

Comment:
Can the return values for all these be enums rather than int?

Reply:
they can indeed. Bad habit, will fix.